### PR TITLE
[CFDP-1073] manual cache flush

### PR DESCRIPTION
**Pull request by asleepingpanda**
### 1. What does this PR do, or why is it needed?
This PR updates the way that we handle caching across the API project to centralize the logic for storing cached data in Redis.

It also introduces a new mutation that allows for Rock to manually flush content cached in Redis by providing an Entity Id and Entity Type Id to Apollos.

### 2. What design trade-offs/decisions were made?
None that I can think of. The only one that seems a bit strange to me is that you will need to pass in your data fetch method into the `Cache.request` as a param. Ideally, it would be build directly into the Request Builder like `this.request().redis(key, expiresIn).get()`, but there's no way for me to do that (from what I can tell unless @vinnyjth can think of a way)

### 3. How do I test this PR?
If you have a Local Redis instance, you can just make a request to a Content Item and check to see if the value got stored. Otherwise, rely on reading through the code and the automated tests

### 4. What automated tests did you write?
* isRequired utility method
* `Cache.request` without all required properties (requestMethod and args.key)
* Simulated a request to `Cache.request` in which the first request is retrieved remotely and the second one hits the `Cache.get` method

### Content Channel Items
* 12 Hour Cache for single Content Items
* 12 Hour Cache for events
* New cache key for Page Builder `{ key: url, value: Content Channel Item Id }`
* 12 Hour Cache for Page Builder url/ids
* Cache Clear resets Single Content Item, All Events if Event, Page Url/Id if applicable

### Groups
* 12 Hour Cache

### Live Streams
* 12 Hour Cache for Related Nodes

### Attribute Matrix
* 1 Hour Cache

## TODO:

- [ ] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [ ] Closes [tag-issues-here]
- [ ] No new warnings in tests, storybook, or apps
- [ ] Assign relevant reviewers
- [ ] Upload GIF(s) of iOS, Android, Web
- [ ] Get one peer approval before merging

## REVIEW:

**Manual QA**

- [ ] QA branch on iOS and ensure it looks/behaves as expected
- [ ] QA branch on Android and ensure it looks/behaves as expected
- [ ] QA branch on Web and ensure it looks/behaves as expected

**Code Review**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

**Comment by asleepingpanda** [comment]
@jordanwade @rkdavidson I set this up and it's not completed all the way yet. I wanted to see if you guys think that this is a good convention before wrapping this up and applying it project wide.

The `Cache.request` is a helper method that will make it _much_ easier to flush and refetch caches for that mutation

File: src/data/redis-cache/data-source.js [hunk 1]
```diff
@@ -1,6 +1,57 @@
...
+
+  /**
+   * For a given request
+   * @param {function}  request         Async method whose return value gets cached.
+   * @param {object}    args
+   * @param {string}    args.key        Key for the value when stored in Redis.
+   * @param {number}    args.expiresIn  The length of time that the cache should be set for.
+   */
+  async request(
+    requestMethod = isRequired('Cache.request', 'requestMethod'),
```
Comment (rkdavidson) [inline src/data/redis-cache/data-source.js:20]: Slick trick for required param enforcement 👍

File: src/data/redis-cache/__tests__/data-source.tests.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import CacheDS from '../data-source';
+
+describe('the matrix item data source', () => {
```
Comment (rkdavidson) [inline src/data/redis-cache/__tests__/data-source.tests.js:3]: Is this a copy/paste hiccup?

```suggestion
describe('the redis cache data source', () => {
```

File: src/data/redis-cache/__tests__/data-source.tests.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    const request = async () => await Cache.request('some random string', requestArgs);
+
+    try {
+      await request();
+    } catch (e) {
+      expect(e).toBeInstanceOf(TypeError);
+    }
+  });
+
+  it('throws a TypeError when an incorrect `key` is passed', async () => {
```
Comment (rkdavidson) [inline src/data/redis-cache/__tests__/data-source.tests.js:46]: I like these simple-but-valuable unit tests. The graceful error handling in `Cache.request` is helpful, too.

**Review by rkdavidson — COMMENTED** [review]
I don't have much off-hand advice on somehow baking this into the `request()` method/class chain, but this should make caching easier to bake into features from the get-go and a bit more reliable/maintainable for content managers, etc.

Looks good 👍
